packages/server/lib/metric/performanceMonitor.js

```diff
@@ -0,0 +1,180 @@
+  DEFAULT_IMA_EVENTS.forEach(eventName => {
+    const isFirstEvent = eventName === START_EVENT;
+    const isLastEvent = eventName === (options.logEvent || END_EVENT);
+
+    // Start listener - runs BEFORE all event handlers
+    const startListener = function PerformanceTrackerStartListener(event) {
+      // Initialize context if needed
+      if (!event.context) {
+        event.context = {};
+      }
+
+      // First event: Create PerformanceTracker
+      if (isFirstEvent) {
+        event.context.perf = new PerformanceTracker({
+          enabled: options.enabled,
+          slowThreshold: options.slowThreshold,
+          includeMetadata: options.includeMetadata,
+          useNativeMarks: options.useNativeMarks,
+          useNativeMeasures: options.useNativeMeasures,
+        });
+
+        // Track request start as a point-in-time event
+        if (options.autoTrackEvents) {
+          event.context.perf.track('request.received', {
+            method: event.req?.method,
+            url: event.req?.url,
+            headers: event.req?.headers
+              ? Object.keys(event.req.headers).length
+              : 0,
+          });
+        }
+      }
+
+      // Start timing this event's handlers
+      if (options.autoTrackEvents && event.context.perf) {
+        event.context.perf.start(eventName);
+      }
+    };
+
+    // End listener - runs AFTER all event handlers
+    const endListener = function PerformanceTrackerEndListener(event) {
+      if (!options.autoTrackEvents || !event.context?.perf) {
+        return;
+      }
+
+      // End timing this event's handlers
+      event.context.perf.end(eventName, {
+        hasError: !!event.error,
+        statusCode: event.res?.statusCode,
+      });
+
+      // Last event: Log report and call completion callback
+      if (isLastEvent) {
+        // Log to console
+        if (options.logToConsole) {
+          event.context.perf.logReport();
+        }
+
+        // Call completion callback with report
+        if (options.onComplete && typeof options.onComplete === 'function') {
+          const report = event.context.perf.getReport();
+
+          if (report) {
+            // Add request context to report
+            report.request = {
+              method: event.req?.method,
+              url: event.req?.url,
+              statusCode: event.res?.statusCode,
+              timestamp: new Date().toISOString(),
+            };
+
+            options.onComplete(report);
+          }
+        }
+      }
+    };
+
+    // Ensure _events Map exists
+    if (!emitter._events) {
+      emitter._events = new Map();
+    }
+
+    // Get existing listeners or create new array
+    const existingListeners = emitter._events.get(eventName) || [];
+
+    // Insert start listener at the beginning (runs FIRST)
+    existingListeners.unshift(startListener);
+
+    // Append end listener at the end (runs LAST)
+    existingListeners.push(endListener);
```

Comment:
Why don't you use `emitter.on`?

Reply:
This is leftover from first implementations where it work differently, will fix :)